plugins/inputs/opensearch_query/opensearch_query.go

```diff
@@ -0,0 +1,197 @@
+	MeasurementName      string          `toml:"measurement_name"`
+	DateField            string          `toml:"date_field"`
+	DateFieldFormat      string          `toml:"date_field_custom_format"`
+	QueryPeriod          config.Duration `toml:"query_period"`
+	FilterQuery          string          `toml:"filter_query"`
+	MetricFields         []string        `toml:"metric_fields"`
+	MetricFunction       string          `toml:"metric_function"`
+	Tags                 []string        `toml:"tags"`
+	IncludeMissingTag    bool            `toml:"include_missing_tag"`
+	MissingTagValue      string          `toml:"missing_tag_value"`
+	mapMetricFields      map[string]string
+	aggregationQueryList []aggregationQueryData
+}
+
+func (*OpensearchQuery) SampleConfig() string {
+	return sampleConfig
+}
+
+// Init the plugin.
+func (o *OpensearchQuery) Init() error {
+	if o.URLs == nil {
+		return fmt.Errorf("opensearch urls is not defined")
+	}
+
+	err := o.connectToOpensearch()
+	if err != nil {
+		return fmt.Errorf("error connecting to opensearch: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(o.Timeout))
+	defer cancel()
+
+	for i, agg := range o.Aggregations {
+		if agg.MeasurementName == "" {
+			return fmt.Errorf("field 'measurement_name' is not set")
+		}
+		if agg.DateField == "" {
+			return fmt.Errorf("field 'date_field' is not set")
+		}
+		err = o.initAggregation(ctx, agg, i)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *OpensearchQuery) initAggregation(ctx context.Context, agg osAggregation, i int) (err error) {
+	// retrieve field mapping and build queries only once
+	agg.mapMetricFields, err = o.getMetricFields(ctx, agg)
+	if err != nil {
+		return fmt.Errorf("not possible to retrieve fields: %v", err.Error())
+	}
+
+	for _, metricField := range agg.MetricFields {
+		if _, ok := agg.mapMetricFields[metricField]; !ok {
+			return fmt.Errorf("metric field '%s' not found on index '%s'", metricField, agg.Index)
+		}
+	}
+
+	err = agg.buildAggregationQuery()
+	if err != nil {
+		return fmt.Errorf("error building aggregation: %s", err)
+	}
+
+	o.Aggregations[i] = agg
+	return nil
+}
+
+func (o *OpensearchQuery) connectToOpensearch() error {
+	var client *opensearch.Client
+	var transport *http.Transport
+
+	if o.InsecureSkipVerify {
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	}
+
+	clientConfig := opensearch.Config{
+		Addresses: o.URLs,
+		Username:  o.Username,
+		Password:  o.Password,
+	}
+
+	if transport != nil {
+		clientConfig.Transport = transport
+	}
+
+	client, err := opensearch.NewClient(clientConfig)
+	if err != nil {
+		return err
+	}
+
+	o.osClient = client
+	return nil
+}
+
+// Gather writes the results of the queries from OpenSearch to the Accumulator.
+func (o *OpensearchQuery) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+
+	err := o.connectToOpensearch()
+	if err != nil {
+		return err
+	}
```

Comment:
A new client is created every gathering cycle unconditionally.

Reply:
Yeah, this is unnecessary behavior from the old library.  It doesn't need to be here at all.